### PR TITLE
Remove set_default_dtype call in jit_metaprogramming_utils.py

### LINKED ISSUE #68972: default dtype is incorrectly set to torch.double when running test_ops.py
Imports into test files should not be setting the default dtype. We should add asserts to the test files after imports that the default dtype is float32. 

@eellison maybe leaking from a jit include? 

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @mruberry

### PR DESCRIPTION
**Pull request by kurtamohler**
Fixes #68972

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

### CODE REVIEW

**Comment by kurtamohler** [comment]
Hey @mruberry, I've made progress so far getting a lot of the tests to work with float32 being the default dtype. However, there are some tests defined in `torch/testing/_internal/common_nn.py` where the reference functions have too much floating point error when they use float32, making them fail. In these cases, I think there are two possible solutions.

One option is to try to modify the reference functions to decrease the error, but that probably isn't always guaranteed to be possible for every case and it would require a good understanding of each function.

The other option is to just go back to using float64 for the tests that fail with due to precision. This is a much easier option. Would that be alright with you?

**Comment by jbschlosser** [comment]
Hey @kurtamohler, I'm in the process of migrating the old `torch/testing/_internal/common_nn.py` module tests to more understandable versions in `test/test_modules.py`. Out of curiosity, have you seen any problems with the tests defined there?

**Comment by kurtamohler** [comment]
> Hey @kurtamohler, I'm in the process of migrating the old `torch/testing/_internal/common_nn.py` module tests to more understandable versions in `test/test_modules.py`. Out of curiosity, have you seen any problems with the tests defined there?

`test/test_modules.py` isn't giving me any problems. The only remaining problem I know about is that some tests in  `torch/testing/_internal/common_nn.py` need to explicitly initialize with doubles to fix gradcheck errors.

**Comment by kurtamohler** [comment]
So, it turns out that if we remove the `set_default_dtype` call from `test_nn.py`, a couple hundred tests in `torch/testing/_internal/common_nn.py` will need to be changed to explicitly be `torch.double`. Is it worth the effort for `test_nn.py`?

Just to be clear, `test_ops.py` and `test_jit.py` require much fewer edits, so it's definitely a good idea to keep going with those test suites.

**Comment by mruberry** [comment]
> So, it turns out that if we remove the `set_default_dtype` call from `test_nn.py`, a couple hundred tests in `torch/testing/_internal/common_nn.py` will need to be changed to explicitly be `torch.double`. Is it worth the effort for `test_nn.py`?
> 
> Just to be clear, `test_ops.py` and `test_jit.py` require much fewer edits, so it's definitely a good idea to keep going with those test suites.

This PR is already great -- what if we just explicitly set the default dtype to double in test_nn.py and file a follow-up issue (with a comment in the code linking the issue) for a "better engineering" TODO. That way we can get this PR merged, and we'll still have an issue to track further test_nn.py changes (which @jbschlosser) is already working on.

How does that sound @kurtamohler? At least that way test_ops.py and test_jit.py won't have the default double dtype, and with asserts to protect them they won't regress again. Plus test_nn.py will have a clear TODO.

**Comment by jbschlosser** [comment]
I'm good with that plan as well. My hope is that eventually the old stuff will just be destroyed and we won't have to make sweeping changes to it in `test_nn.py`.

**Comment by kurtamohler** [comment]
@mruberry, this is ready for another review

**Review by eellison — APPROVED** [review]
Wow, nice!! LGTM

**Review by mruberry — APPROVED** [review]
Thanks for the update, @kurtamohler! Sorry this slipped through the crackers in early January!

**Comment by mruberry** [comment]
Internal CI is failing the assertion `assert torch.get_default_dtype() == torch.float32`. Which is unfortunate. 

We have a mechanism to tell if we're running in internal CI or not, 

https://github.com/pytorch/pytorch/blob/b36b11cbc1f11fa805894927ca667443f3618a34/torch/testing/_internal/common_utils.py#L74

I suggest we just guard the assertion on `IS_SANDCASTLE` and `IS_FBCODE` with a comment that the assertion fails when run in Sandcastle. That way the OSS tests will be run in the correct dtype and the internal tests can continue doing whatever for the moment. 

Does that sound OK, @kurtamohler?

File: test/test_jit.py [hunk 1]
```diff
@@ -143,6 +144,10 @@
 import warnings
 import zipfile
 
+# This assertion fails when run in internal CI
+if not IS_SANDCASTLE and not IS_FBCODE:
```
Comment (kurtamohler) [inline test/test_jit.py:148]: @mruberry, just making sure, this is what we want, right?

Comment (mruberry) [reply test/test_jit.py:148]: Yep

**Comment by mruberry** [comment]
FYI this is just going to require a couple internal tweaks to land -- no action needed

**Comment by kurtamohler** [comment]
@mruberry any word on the internal issues?

**Comment by kurtamohler** [comment]
@mruberry, did you get a chance to look into the internal issues with this one?

I'll look into the merge conflicts soon

**Comment by kit1980** [comment]
@kurtamohler Do you still want to merge this? If yes, please resolve merge conflicts and I can take over Meta-internal verification.

**Comment by kurtamohler** [comment]
@kit1980, thanks for the ping. Yes I would like to get this merged. I will resolve the conflicts

**Comment by kurtamohler** [comment]
@kit1980 , I think this is ready for Meta-internal verification. The remaining CI failure is upstream

**Comment by kit1980** [comment]
@kurtamohler I see one relevant Meta-internal test failure:
```
RuntimeError: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript (most recent call last):
  File "<string>", line 3, in the_method

def the_method(i0, i1, i2):
    return torch.nn.functional.bilinear(i0, i1, i2)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
RuntimeError: expected scalar type Double but found Float
```

This is a generated test. Not sure why we don't have similar on GitHub or if it's even a good test.
Is this something you can address on your end or I can dig deeper?

**Comment by kurtamohler** [comment]
Thanks for the update @kit1980. I'm guessing that's something that will have to be fixed internally. It seems like the arguments to bilinear are not all the same type. For instance, I get the same error with this:

```python
>>> torch.nn.functional.bilinear(torch.randn(3).double(), torch.randn(5), torch.randn(1, 3, 5))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: expected scalar type Double but found Float
```

**Comment by kit1980** [comment]
@kurtamohler Actually lots of tests failing on GitHub with what looks like the same cause:
https://github.com/pytorch/pytorch/actions/runs/3914201447/jobs/6809694610

    RuntimeError: mixed dtype (CPU): expect parameter to have scalar type of Float

    RuntimeError: expected scalar type Float but found Double

**Comment by kurtamohler** [comment]
True, it's possible that fixing those errors will fix the internal ones. I've figured out how to reproduce some of the errors in my environment, now I'll work on fixing them

**Pull request closed without merging**